---
title: Differential Gene Expression for Maier Bulk RNA-Seq
subtitle: Compares Sarcoidosis to Control
author: |
 | Project: `r CIDAtools::ProjectName()`
 |
 | Analyst: `r CIDAtools::ProjectAnalyst()`
 |
 | Investigator(s): `r CIDAtools::ProjectPI()`
 |
 | Report generated: `r paste(format(Sys.Date(), '%B %d, %Y'))`
output:
  html_document:
    code_folding: show
    highlight: espresso
    number_sections: yes
    theme: yeti
    toc: yes
    toc_float: yes
---

```{r, echo=FALSE, warning=F, out.width='70%', fig.show='hold'}
logo <- system.file("figures", "sph_cida_wm_blk.png", package="CIDAtools")
knitr::include_graphics(logo)
```

---

```{r setup, include=F}
knitr::opts_chunk$set(echo=F, warning=F, message=F, fig.align = "center")
library(tidyverse)
library(knitr)
library(kableExtra)
library(openxlsx)
library(edgeR)
library(DESeq2)
library(plotly)
library(tableone)
library(DT)
library(enrichR)
```

```{r}
`%notin%` <- Negate(`%in%`)

options(future.globals.maxSize = 100*1000 * 1024^2)

source("processingData.R")
source("XX_formattingFunctions.R")
source("XX_enrichmentFunctions.R")

nSmallGroup <- 13

dataDir <- "P:/BRANCHES/NJHMoore_CBD_Sarcoidosis/maier_bulkRNASeq_paperAnalysis/DataRaw/"
outDir  <- "P:/BRANCHES/NJHMoore_CBD_Sarcoidosis/maier_bulkRNASeq_paperAnalysis/"
ntab <- 1
nfig <- 1

outDat <- getDataStarted(dataDir, nSmallGroup, outliers = T)
counts.filtered <- outDat[[1]]
all_md <- outDat[[2]]
fdata <- outDat[[3]]
genes <- outDat[[4]]

# Create DESeq2 Dataset and Model
dds <- DESeqDataSetFromMatrix(countData = counts.filtered, colData = all_md,
                              design = ~ Birth.sex + Dx, rowData = fdata)
dds <- DESeq(dds)

# Simple DESeq2 Differential Expression Analysis #####

# Get results for the comparision of the groups
Sarc_v_Control <- results(dds)
Sarc_v_Control <- Sarc_v_Control[order(Sarc_v_Control$padj),]
# Add in the gene symbols
Sarc_v_Control$gene <- genes[rownames(Sarc_v_Control)]
```

# Background

Code sourced from Arunangshu Sarkar. This report was generated to recreate results located in html reports previously created by Arunangshu Sarkar. The results found in the manuscript draft are replicated here. Source code for processing the data is listed in the appendix. 

# Methods

Genes were only included if there were at least 13 samples (the number of samples in the smallest group) with at least 1 counts per million (CPM). Samples were evaluated for outliers using principle component analysis (PCA), these samples were removed from analysis, and details on the PCA are located in *Preliminary_Sarc_withOutliers_redo.Rmd*. 

A Wald test was used to compare Sarcoidosis diagnosis to Control. Sex was included in the DESeq2 Model. To correct for multiple testing the Benjamini and Hochberg False Discovery Rate (FDR) was used. 

# Results 

There are 3 samples removed (39810, 33042, 16092). There are `r ncol(counts.filtered)` subjects and `r prettyNum(nrow(counts.filtered), big.mark = ",")` genes included in the analysis. 

**Table `r ntab`. Population Demographics**

```{r}
# double check rows of MD and columns of counts match / are ordered correctly
#table(rownames(all_md)==colnames(counts))
t1 <- CreateTableOne(vars = c('Age.at.sampling', 'Birth.sex', 'Race', 'Hispanic', 'Smoking.Status'), 
                     strata = 'Phenotype', data = all_md)
t1.print <- print(t1, nonnormal = 'Age.at.sampling', exact = c('Birth.sex', 'Race', 'Hispanic', 'Smoking.Status'),  printToggle = FALSE)

t1.print[which(t1.print == " 1.000")] <- "> 0.99"

t1.print %>% kable(align = "c", col.names = c("Control", "Non-progressive", "Progressive", "p-value", "Test Used")) %>%
  kable_paper("hover", full_width = T)
```

## Differential Gene Expression Results

### Sarcoidosis vs Control Comparision 

```{r}
Sarc_v_Control_df <- formatDat(Sarc_v_Control)
```

`r nrow(Sarc_v_Control_df %>% filter(padj < 0.05))` genes significantly differed between the Sarcoidosis and Control samples (FDR < 0.05). There were an additional `r nrow(Sarc_v_Control_df %>% filter(padj >= 0.05 & padj < 0.1))` genes nominally differ (FDR < 0.1). 

**Table `r ntab`. Sarcoidosis v Control Results**

```{r}
ntab <- ntab + 1
getDEResultsTab(Sarc_v_Control_df)
```

**Figure `r nfig`. Volcano Plot**

```{r}
nfig <- nfig + 1
ggplotly(plotVolcano(Sarc_v_Control_df))
```

## Enrichment Analysis Results

Pathways or ontologies were only included below in results table if it meet the significance threshold of 0.05 for the adjusted p-value. 

### Sarcoidosis vs Control Comparision 

**Table `r ntab`. Significant Pathways and Ontologies for Up-regulated Genes**

```{r, include = F}
setEnrichrSite("Enrichr") # Human genes
dbs <- c("BioPlanet_2019", "Reactome_2016", "GO_Biological_Process_2021", "GO_Molecular_Function_2018", "GO_Cellular_Component_2018", "ARCHS4_Tissues", "KEGG_2019_Human", "MSigDB_Hallmark_2020", "WikiPathways_2019_Human")

Sarc_v_Control_enrich <- runEnrichment(Sarc_v_Control)
```

```{r}
ntab <- ntab + 1
getEnResutltsTab(Sarc_v_Control_enrich[[1]])
```

**Table `r ntab`. Significant Pathways and Ontologies for Down-regulated Genes**

```{r}
ntab <- ntab + 1
getEnResutltsTab(Sarc_v_Control_enrich[[2]])
```

# Appendix

**Source Code: ProcessingData.R**

```{r}
readLines('processingData.R')
```

**Session Information**

```{r}
sessionInfo()
```

<!-- footer -->

---

```{r, echo=FALSE, out.width='70%', fig.show='hold'}
knitr::include_graphics(logo)
```

